### PR TITLE
[issue - 1894] - Run sonar during install

### PR DESCRIPTION
**Pull request by pmbrull**
### Describe your changes :

This PR fixes https://github.com/open-metadata/OpenMetadata/issues/1894

The goal of this PR is to add a Static Code Analysis step during the `verify` phase. When running `mvn install`, this will trigger the `SCA` profile if we are in the `main` branch. This will speed up PRs and local checks while allowing us to have a continuous analysis of the production code.

As this is a multi-module project, we needed to add the core specification in the parent pom. The module-specific properties are added to each child. In our case, this just means so far the `catalog-rest-service`.

We have updated the CI job to have the following shape:

```
mvn -Dbranch=${{ github.ref_name }} -Dsonar.login=${{ secrets.SONAR_TOKEN }} clean package
```

As we are passing the branch, `mvn` will skip the `SCA` for any branch but `main`. We are also providing the token saved under secrets.

I have tested this running sonar on a Docker locally and these are the results:

image

Note how the process properly adds test & coverage information for the backend code.

In another PR we will prepare the same setup for Python. I believe it would be best to prepare a separate project in SonarCloud for the ingestion.

Without providing any explicit exclusions, just by setting the language to `java`, these are the files detected by the analysis:

image

We might need to disable the automatic analysis on SonarCloud and have a proper manual build with a more controlled environment.

Let me know your thoughts! Thanks 🚀 

> OBS: I commented out the webhook tests locally to make the code run. CI will break by `EntityResourceTest:249`

#
### Type of change :

- [ ] Bug fix
- [ ] Improvement
- [x] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/open-source-community/developer) document.
- [ ] I have performed a self-review of my own. 
- [ ] I have tagged my reviewers below.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

#
### Reviewers

@harshach @mithmatt

### CODE REVIEW

**Comment by pmbrull** [comment]
It will also be interesting to update the properties and filter out the generated code etc 🤔

File: pom.xml [hunk 1]
```diff
@@ -420,14 +430,64 @@
   </repositories>
 
   <profiles>
+
+    <profile>	<!-- run SCA for master branch: -Dbranch=master -->
+      <id>static-code-analysis</id>
+      <activation>
+        <property>
```
Comment (mithmatt) [inline pom.xml:437]: Should we remove the `branch=main` property here and enable devs to run this on any branch (especially code coverage report in local laptop/desktop dev environments, without changing pom)?

We can always do `[ "${{ github.ref_name }}" = "main" ] && mvn -P static-code-analysis clean package` to activate this profile for the CI.

**Comment by mithmatt** [comment]
> It is interesting as well to force a coverage % to ensure maintainability.

@pmbrull @harshach 

If we have to force a coverage %, we need to think a bit further:

- Target coverage % should be determined carefully so that it does not hamper developer productivity and subsequently cause less engagement from the community
- To keep coverage % up, then each PR should report coverage increase or decrease % - this implicitly signals the contributor to increase coverage and surfaces up to reviewers if contributor has done due diligence writing tests

**Comment by harshach** [comment]
@pmbrull @mithmatt For now we can keep this as an indicator rather than enforcer such that we don't decrease the velocity as @mithmatt pointed out but as a community we know whats our coverage looks like and we can continue to increase that. Once we get into 1.0 release we can target for a higher coverage and make it a enforcer i.e a PR should contain tests etc..

**Comment by pmbrull** [comment]
Sure thing! Just wanted to pop the discussion, makes sense to delay this to 1.0 :)

Thanks!

**Comment by mithmatt** [comment]
+1 to NOT enforcing for now (and letting me sneak in more code without worrying about tests 😃 )

**Comment by pmbrull** [comment]
Might have seen some vulnerabilities with your name @mithmatt :whistle: heh

File: pom.xml (hunk 1 above)
Comment (pmbrull) [reply pom.xml]: agree, took a spin to the CI workflow to work with PRs and main pushes. With PRs, we will also pass specific information that sonar can use to build the analysis for the changes

**Comment by pmbrull** [comment]
Hi, @harshach I am trying to set the coverage information for the PRs as well. Could you please modify the SonarCloud settings in `Administration > Analysis Method > Set SonarCloud Automatic Analysis to off`? let's see if then I can pass the proper information during the CI. Thanks!

**Comment by harshach** [comment]
@pmbrull made that change. can you please check

**Comment by pmbrull** [comment]
Hi @harshach @mithmatt, I believe we have the correct setup now.

To the question, why are we seeing Sonar showing up all those bugs in the code just for a change in the CI? I believe that this is because the automatic analysis we did so far in the project did not pick up any Java code. With our configurations, now it has been able to discover it (and the results are aligned with the screenshot of my local execution).

In order not to show all the builds as failed and just have sonar as an informational source, we should update the quality gate to be totally permissive. In order to do so:
1. Under the organization settings > Quality Gates, create a new quality gate with a single condition, e.g. `Blocker issues is greater than 0`.
2. Now, on OpenMetadata project > Administration > Quality Gate > Select the new gate

This way we will have all the analysis and historical information without forcing contributors to fix incoming code. Bit by bit we can then add new conditions to the created Quality Gate.

Let me know what you think, happy to keep discussing this! Also, as the project has multiple modules with different languages it can be a bit tricky, so we might need to further iterate on the properties to include/exclude parts of the code.

What I believe would make sense are different projects on sonar, each with its own language setting to tackle issues separately. Otherwise, it can feel a bit overwhelming.

Thanks 🎄

**Comment by harshach** [comment]
image

**Comment by harshach** [comment]
@pmbrull updated the quality gate as you suggested. It does look green for the main branch not sure why this PR's check is failing

**Comment by pmbrull** [comment]
We might need to rerun the CI for it to update

**Comment by pmbrull** [comment]
Hi @harshach my bad. In the example I said 0 blocker issue, but we have some of them in there.

For the sake of setting a dummy validation and keep updating and monitoring it further, could we have a higher number? E.g. 999?

I hope that should settle this. Thanks 🙏

**Comment by pmbrull** [comment]
Thanks, @harshach. It looks like it is picking up the new code and passing the checks.

Please, let me know if there's something else we should do here or if we are good to go 🚀

**Review by mithmatt — APPROVED** [review]

**Review by harshach — APPROVED** [review]

**Comment by harshach** [comment]
Thanks @pmbrull . Looks great

**Pull request merged**